tiledb/api/src/array/enumeration.rs

```diff
@@ -0,0 +1,502 @@
+        });
+        write!(f, "{}", json)
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Enumeration<'c2>> for Enumeration<'c1> {
+    fn eq(&self, other: &Enumeration<'c2>) -> bool {
+        let names_match = match (self.name(), other.name()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !names_match {
+            return false;
+        }
+
+        let types_match = match (self.datatype(), other.datatype()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !types_match {
+            return false;
+        }
+
+        let cell_val_num_match =
+            match (self.cell_val_num(), other.cell_val_num()) {
+                (Ok(mine), Ok(theirs)) => mine == theirs,
+                _ => false,
+            };
+        if !cell_val_num_match {
+            return false;
+        }
+
+        let ordered_match = match (self.ordered(), other.ordered()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !ordered_match {
+            return false;
+        }
+
+        // TODO: Match data and offsets
+
+        true
+    }
+}
+
+pub struct Builder<'ctx> {
+    context: &'ctx Context,
+    name: String,
+    dtype: Datatype,
+    cell_val_num: u32,
+    ordered: bool,
+}
+
+impl<'ctx> Builder<'ctx> {
+    pub fn new(context: &'ctx Context, name: &str, dtype: Datatype) -> Self {
+        Builder {
+            context,
+            name: name.to_owned(),
+            dtype,
+            cell_val_num: 1,
+            ordered: false,
+        }
+    }
+
+    pub fn cell_val_num(self, cell_val_num: u32) -> Self {
+        Self {
+            cell_val_num,
+            ..self
+        }
+    }
+
+    pub fn var_sized(self) -> Self {
+        Self {
+            cell_val_num: u32::MAX,
+            ..self
+        }
+    }
+
+    pub fn ordered(self, ordered: bool) -> Self {
+        Self { ordered, ..self }
+    }
+
+    pub fn build(self, column: Column) -> TileDBResult<Enumeration<'ctx>> {
+        let mut c_enmr: *mut ffi::tiledb_enumeration_t = out_ptr!();
+        let name_bytes = self.name.as_bytes();
+        let c_name = cstring!(name_bytes);
+        let c_dtype = self.dtype.capi_enum();
+
+        // Rust semantics require that slice pointers aren't nullptr so that
+        // nullptr can be used to distinguish between Some and None. The stdlib
+        // empty slices all appear to return 0x1 which is mentioned in the docs
+        // as a valid strategy. For our situation, we just use a zero length to
+        // indicate when we should pass nullptr.
+        let offsets_ptr = if column.offsets().is_empty() {
+            std::ptr::null_mut()
+        } else {
+            column.offsets().as_ptr()
+        };
+
+        // An important note here is that the Enumeration allocator copies the
+        // contents of data of offsets rather than assumes ownership. That
```

Comment:
of data *and* offsets? is that what you mean?

Reply:
Totes, will fix.